flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/sinks/TableSinkUtils.scala

```diff
@@ -189,13 +190,45 @@ object TableSinkUtils {
       queryLogicalType,
       withChangeFlag)
     if (LogicalTypeChecks.isCompositeType(requestedOutputType.getLogicalType)) {
-      DataTypeUtils.expandCompositeTypeToSchema(requestedOutputType)
+      fromDataTypeToTypeInfo(requestedOutputType) match {
+        case pj: PojoTypeInfo[_] => expandPojoTypeToSchema(pj, queryLogicalType)
+        case _ => DataTypeUtils.expandCompositeTypeToSchema(requestedOutputType)
+      }
     } else {
       // atomic type
       TableSchema.builder().field("f0", requestedOutputType).build()
     }
   }
 
+  /**
```

Comment:
BTW, don't reorder the nested data in `SinkCodeGenerator`?

Reply:
Unfortunately, we can't reorder the nested data in `SinkCodeGenerator` for now. Because the `ExprCodeGenerator` only accetps the top-level `mapping`. If we want to support reorder nested data, `ExprCodeGenerator` requires a major refactoring. I will leave that as a future work and it is not a big requirement as of right now (it's hard to construct a nested named row for now).

Reply:
Besides, we have to keep the nested fields reordering in `TableSinkUtils` to make the simplest case `StreamTableEnvironmentITCase#testToAppendStreamWithPojoType` passed (no nested reordering required when code generation).